docs/src/modules/ROOT/pages/quickstart/school-timetabling/school-timetabling-constraints.adoc

```diff
@@ -122,5 +172,150 @@ public class TimetableConstraintProvider implements ConstraintProvider {
 
 }
 ----
+--
+
+Kotlin::
++
+--
+Create a `src/main/kotlin/org/acme/schooltimetabling/solver/TimetableConstraintProvider.kt` class:
+
+[source,kotlin]
+----
+package org.acme.schooltimetabling.solver
+
+import ai.timefold.solver.core.api.score.buildin.hardsoft.HardSoftScore
+import ai.timefold.solver.core.api.score.stream.Constraint
+import ai.timefold.solver.core.api.score.stream.ConstraintFactory
+import ai.timefold.solver.core.api.score.stream.ConstraintProvider
+import ai.timefold.solver.core.api.score.stream.Joiners
+import org.acme.schooltimetabling.domain.Lesson
+import org.acme.schooltimetabling.solver.justifications.RoomConflictJustification
+import org.acme.schooltimetabling.solver.justifications.StudentGroupConflictJustification
+import org.acme.schooltimetabling.solver.justifications.StudentGroupSubjectVarietyJustification
+import org.acme.schooltimetabling.solver.justifications.TeacherConflictJustification
+import org.acme.schooltimetabling.solver.justifications.TeacherRoomStabilityJustification
+import org.acme.schooltimetabling.solver.justifications.TeacherTimeEfficiencyJustification
+import java.time.Duration
+
+class TimeTableConstraintProvider : ConstraintProvider {
+
+    override fun defineConstraints(constraintFactory: ConstraintFactory): Array<Constraint>? {
+        return arrayOf(
+            // Hard constraints
+            roomConflict(constraintFactory),
+            teacherConflict(constraintFactory),
+            studentGroupConflict(constraintFactory),
+            // Soft constraints
+            teacherRoomStability(constraintFactory),
+            teacherTimeEfficiency(constraintFactory),
+            studentGroupSubjectVariety(constraintFactory)
+        )
+    }
+
+    fun roomConflict(constraintFactory: ConstraintFactory): Constraint {
+        // A room can accommodate at most one lesson at the same time.
+        return constraintFactory
+            // Select each pair of 2 different lessons ...
+            .forEachUniquePair(
+                Lesson::class.java,
+                // ... in the same timeslot ...
+                Joiners.equal(Lesson::timeslot),
+                // ... in the same room ...
+                Joiners.equal(Lesson::room)
+            )
+            // ... and penalize each pair with a hard weight.
+            .penalize(HardSoftScore.ONE_HARD)
+            .justifyWith({ lesson1: Lesson, lesson2: Lesson?, score: HardSoftScore? ->
```

Comment:
parentheses can be removed as the lambda is the only parameter.
all parameters are not-null in this case, aren't they? removing lesson2's "?" will eliminate the need for not-null casting ("!!") in the next line